File: plot_veldiagram.m
function plot_veldiagram
global SA plot_setting param

% What is currently selected
cols='bgr';

%% create figure and axes
ax = findobj('Tag','VelPlot');
if isempty(ax)
    fig = figure('Position', param.TablePosition, ...
        'Tag','VelPlot_figure', 'PaperType','a4', 'PaperOrientation','landscape',...
        'NumberTitle','off', 'Name','Velocity Plot');
    
    addToolbarExplorationButtons(fig)
    ax = axes('Position',[0.1    0.1    0.8    0.8], 'Tag','VelPlot',  'NextPlot', 'add');
    hold(ax, 'on')
else
   delete(findobj('Tag','myline_VelPlot'));
end

%% Create and plot velocity
legtxt={};
legi=0;
for si = SA.Current_si
    [b1b2b3, b1i, b2i, b3i] = str2num_B1B2B3(SA.Table.B1B2B3{si});
    for bi =1:3
        ball(bi).vx = [[diff(SA.Shot(si).Route0(bi).x) ./ diff(SA.Shot(si).Route0(bi).t)]; 0];
        ball(bi).vy = [[diff(SA.Shot(si).Route0(bi).y) ./ diff(SA.Shot(si).Route0(bi).t)]; 0];
        ball(bi).t = SA.Shot(si).Route0(bi).t;
        if b1b2b3(bi) ~= 1
            ball(bi).t = [0; ball(bi).t(2); ball(bi).t(2:end)];
            ball(bi).vx = [0; 0; ball(bi).vx(2:end)];
            ball(bi).vy = [0; 0; ball(bi).vy(2:end)];
        end
        
        ball(bi).v = sqrt(ball(bi).vx.^2 + ball(bi).vy.^2)/1000;
    end
    
    
    for bi =1:3
        plot(ax, ball(b1b2b3(bi)).t, ball(b1b2b3(bi)).v, ['-',cols(bi)],'Tag','myline_VelPlot')
        legi=legi+1;
        legtxt{legi} = ['Shot ',num2str(si),' B',num2str(bi)];
    end
    
end
grid(ax, 'on')
title(ax, 'Balls Speeds')
xlabel(ax,'time in s')
ylabel(ax, 'velocity in m/s')
% legend(ax, legtxt);

File: plot_veldiagram.py
import matplotlib.pyplot as plt
import numpy as np

def plot_veldiagram(SA, param):
    cols = ['b', 'g', 'r']

    fig, ax = plt.subplots()
    fig.set_size_inches(10, 6)
    fig.canvas.manager.set_window_title("Velocity Plot")

    legtxt = []

    for si in SA['Current_si']:
        b1b2b3, b1i, b2i, b3i = str2num_B1B2B3(SA['Table']['B1B2B3'][si])
        ball = []

        for bi in range(3):
            vx = np.diff(SA['Shot'][si]['Route0'][bi]['x'], append=0) / np.diff(SA['Shot'][si]['Route0'][bi]['t'], append=1)
            vy = np.diff(SA['Shot'][si]['Route0'][bi]['y'], append=0) / np.diff(SA['Shot'][si]['Route0'][bi]['t'], append=1)
            t = SA['Shot'][si]['Route0'][bi]['t']

            if b1b2b3[bi] != 1:
                t = np.insert(t, 1, t[1])
                vx = np.insert(vx, 1, 0)
                vy = np.insert(vy, 1, 0)

            v = np.sqrt(vx**2 + vy**2) / 1000
            ball.append({'t': t, 'v': v})

        for bi in range(3):
            ax.plot(ball[b1b2b3[bi] - 1]['t'], ball[b1b2b3[bi] - 1]['v'], f'-{cols[bi]}', label=f'Shot {si} B{bi + 1}')

    ax.grid(True)
    ax.set_title("Balls Speeds")
    ax.set_xlabel("Time in s")
    ax.set_ylabel("Velocity in m/s")
    ax.legend()

    plt.show()